### PR TITLE
Fixes DTD completion

### LINKED ISSUE #232: Completion for DTD <!ELEMENT, <!ATTRIBUTE,
Support for completion 

 * for DTD elements 

`<! <- Ctrl+Space shows <!ELEMENTS <!ATTRIBUTE, <!ENTITY`

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #232

### CODE REVIEW

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMDocument.java [hunk 1]
```diff
@@ -760,6 +761,50 @@ public boolean isDTD() {
...
+	/**
+	 * Returns true if 'offset' is within an internal DOCTYPE dtd.
+	 * Else false.
+	 * @param offset
+	 * @return 
+	 */
+	public boolean isWithinInternalDTD(int offset) {
+		DOMDocumentType doctype = this.getDoctype();
+		if(doctype != null) {
+			if(doctype.internalSubset != null) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/dom/DOMDocument.java]: combine both ifs

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 2]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
+	}
+	private void collectInsideDTDContent(CompletionRequest request, CompletionResponse response, boolean isContent) {
 		// Insert DTD Element Declaration
 		// see https://www.w3.org/TR/REC-xml/#dt-eldecl
 		CompletionItem elementDecl = new CompletionItem();
 		elementDecl.setLabel("Insert DTD Element declaration");
 		elementDecl.setKind(CompletionItemKind.EnumMember);
-		elementDecl.setFilterText("<!ELEMENT");
+		elementDecl.setFilterText("<!ELEMENT ");
 		elementDecl.setInsertTextFormat(InsertTextFormat.Snippet);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 3]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
+				if(editRange == null) {
+					editRange = getReplaceRange(node.getStart(), node.getEnd(), request);
+				}
+			}
 		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ELEMENT completion.", e);
+			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD completion.", e);
 		}
+				
+		elementDecl.setTextEdit(new TextEdit(editRange, "<!ELEMENT ${0:element-name} (${1:#PCDATA})>"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client. 

Also, for some reason vscode selects placeholder 1 instead of 0. Check if using 1 and 2 instead works better

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 4]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
 		response.addCompletionItem(elementDecl);
 
 		// Insert DTD AttrList Declaration
 		// see https://www.w3.org/TR/REC-xml/#attdecls
 		CompletionItem attrListDecl = new CompletionItem();
 		attrListDecl.setLabel("Insert DTD Attributes list declaration");
 		attrListDecl.setKind(CompletionItemKind.EnumMember);
-		attrListDecl.setFilterText("<!ATTLIST");
+		attrListDecl.setFilterText("<!ATTLIST ");
 		attrListDecl.setInsertTextFormat(InsertTextFormat.Snippet);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 5]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			attrListDecl.setTextEdit(new TextEdit(editRange,
-					"<!ATTLIST ${0:element-name} ${1:attribute-name} ${2:ID} ${3:#REQUIRED} >"));
-			attrListDecl.setDocumentation("<!ATTLIST element-name attribute-name ID #REQUIRED >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ATTLIST completion.", e);
-		}
+		
+		attrListDecl.setTextEdit(new TextEdit(editRange,
+				"<!ATTLIST ${0:element-name} ${1:attribute-name} ${2:ID} ${3:#REQUIRED} >"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 6]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
 		response.addCompletionItem(attrListDecl);
 
 		// Insert Internal DTD Entity Declaration
 		// see https://www.w3.org/TR/REC-xml/#dt-entdecl
 		CompletionItem internalEntity = new CompletionItem();
 		internalEntity.setLabel("Insert Internal DTD Entity declaration");
 		internalEntity.setKind(CompletionItemKind.EnumMember);
-		internalEntity.setFilterText("<!ENTITY");
+		internalEntity.setFilterText("<!ENTITY ");
 		internalEntity.setInsertTextFormat(InsertTextFormat.Snippet);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 7]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
 		startOffset = request.getOffset();
-		try {
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			internalEntity.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} \"${1:entity-value}\" >"));
-			internalEntity.setDocumentation("<!ENTITY entity-name \"entity-value\" >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ENTITY completion.", e);
-		}
+		
+		internalEntity.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} \"${1:entity-value}\" >"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 8]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
 		response.addCompletionItem(internalEntity);
 
 		// Insert External DTD Entity Declaration
 		// see https://www.w3.org/TR/REC-xml/#dt-entdecl
 		CompletionItem externalEntity = new CompletionItem();
 		externalEntity.setLabel("Insert External DTD Entity declaration");
 		externalEntity.setKind(CompletionItemKind.EnumMember);
-		externalEntity.setFilterText("<!ENTITY");
+		externalEntity.setFilterText("<!ENTITY ");
 		externalEntity.setInsertTextFormat(InsertTextFormat.Snippet);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 9]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			externalEntity
-					.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} SYSTEM \"${1:entity-value}\" >"));
-			externalEntity.setDocumentation("<!ENTITY entity-name SYSTEM \"entity-value\" >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ENTITY completion.", e);
-		}
+		
+		externalEntity
+				.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} SYSTEM \"${1:entity-value}\" >"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: need to check if snippets supported by client

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java (hunk 5 above)
Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: remove trailing space before closing bracket

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java (hunk 7 above)
Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: remove trailing space before closing bracket

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 10]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-		try {
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			internalEntity.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} \"${1:entity-value}\" >"));
-			internalEntity.setDocumentation("<!ENTITY entity-name \"entity-value\" >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ENTITY completion.", e);
-		}
+		
+		internalEntity.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} \"${1:entity-value}\" >"));
+		internalEntity.setDocumentation("<!ENTITY entity-name \"entity-value\" >");
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: remove trailing space before closing bracket

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java (hunk 9 above)
Comment (fbricon) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: remove trailing space before closing bracket

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 11]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-			externalEntity
-					.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} SYSTEM \"${1:entity-value}\" >"));
-			externalEntity.setDocumentation("<!ENTITY entity-name SYSTEM \"entity-value\" >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ENTITY completion.", e);
-		}
+		
+		externalEntity
+				.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} SYSTEM \"${1:entity-value}\" >"));
+		externalEntity.setDocumentation("<!ENTITY entity-name SYSTEM \"entity-value\" >");
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: remove trailing space before closing bracket

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 12]
```diff
@@ -691,74 +704,82 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-			attrListDecl.setTextEdit(new TextEdit(editRange,
-					"<!ATTLIST ${0:element-name} ${1:attribute-name} ${2:ID} ${3:#REQUIRED} >"));
-			attrListDecl.setDocumentation("<!ATTLIST element-name attribute-name ID #REQUIRED >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ATTLIST completion.", e);
-		}
+		
+		attrListDecl.setTextEdit(new TextEdit(editRange,
+				"<!ATTLIST ${0:element-name} ${1:attribute-name} ${2:ID} ${3:#REQUIRED} >"));
+		attrListDecl.setDocumentation("<!ATTLIST element-name attribute-name ID #REQUIRED >");
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: remove trailing space before closing bracket

**Comment by fbricon** [comment]
So according to the LSP spec : 

> `$0` defines the final tab stop, it defaults to the end of the snippet

All snippets should be revisited

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 13]
```diff
@@ -691,74 +704,83 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
+				if(editRange == null) {
+					editRange = getReplaceRange(node.getStart(), node.getEnd(), request);
+				}
+			}
 		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ELEMENT completion.", e);
+			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD completion.", e);
 		}
+				
+		elementDecl.setTextEdit(new TextEdit(editRange, "<!ELEMENT ${1:element-name} (${0:#PCDATA})>"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: you can't use placeholders if snippets are not supported. You need to check insertFormat.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 14]
```diff
@@ -691,74 +704,83 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			attrListDecl.setTextEdit(new TextEdit(editRange,
-					"<!ATTLIST ${0:element-name} ${1:attribute-name} ${2:ID} ${3:#REQUIRED} >"));
-			attrListDecl.setDocumentation("<!ATTLIST element-name attribute-name ID #REQUIRED >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ATTLIST completion.", e);
-		}
+		
+		attrListDecl.setTextEdit(new TextEdit(editRange,
+				"<!ATTLIST ${1:element-name} ${2:attribute-name} ${3:ID} ${0:#REQUIRED}>"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: you can't use placeholders if snippets are not supported. You need to check insertFormat.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 15]
```diff
@@ -691,74 +704,83 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
 		startOffset = request.getOffset();
-		try {
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			internalEntity.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} \"${1:entity-value}\" >"));
-			internalEntity.setDocumentation("<!ENTITY entity-name \"entity-value\" >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ENTITY completion.", e);
-		}
+		
+		internalEntity.setTextEdit(new TextEdit(editRange, "<!ENTITY ${1:entity-name} \"${0:entity-value}\">"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: you can't use placeholders if snippets are not supported. You need to check insertFormat.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java [hunk 16]
```diff
@@ -691,74 +704,83 @@ private void collectAttributeValueSuggestions(int valueStart, int valueEnd, Comp
...
-			Range editRange = getReplaceRange(startOffset, startOffset, request);
-			externalEntity
-					.setTextEdit(new TextEdit(editRange, "<!ENTITY ${0:entity-name} SYSTEM \"${1:entity-value}\" >"));
-			externalEntity.setDocumentation("<!ENTITY entity-name SYSTEM \"entity-value\" >");
-		} catch (BadLocationException e) {
-			LOGGER.log(Level.SEVERE, "While performing getReplaceRange for DTD !ENTITY completion.", e);
-		}
+		
+		externalEntity
+				.setTextEdit(new TextEdit(editRange, "<!ENTITY ${1:entity-name} SYSTEM \"${0:entity-value}\">"));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLCompletions.java]: you can't use placeholders if snippets are not supported. You need to check insertFormat.

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/DTDCompletionExtensionsTest.java [hunk 17]
```diff
@@ -83,6 +83,50 @@ public void externalDTDCompletionAttribute() throws BadLocationException {
...
+				"	|\r\n" + //
+				"]>\r\n" + //
+				"<Folks>\r\n" + //
+				"	" + //
+				"</Folks>";
+		XMLAssert.testCompletionFor(xml, 4, c("Insert DTD Element declaration", te(3, 1, 3, 1, "<!ELEMENT ${1:element-name} (${0:#PCDATA})>"), "<!ELEMENT ")
+		,c("Insert Internal DTD Entity declaration", te(3, 1, 3, 1, "<!ENTITY ${1:entity-name} \"${0:entity-value}\">"), "<!ENTITY ")
+		,c("Insert DTD Attributes list declaration", te(3, 1, 3, 1, "<!ATTLIST ${1:element-name} ${2:attribute-name} ${3:ID} ${0:#REQUIRED}>"), "<!ATTLIST ")
+		,c("Insert External DTD Entity declaration", te(3, 1, 3, 1, "<!ENTITY ${1:entity-name} SYSTEM \"${0:entity-value}\">"), "<!ENTITY "));
+	}
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/DTDCompletionExtensionsTest.java]: you can also test completionItems if client doesn't support snippets

**Comment by NikolasKomonen** [comment]
@fbricon updated snippets not supported case

**Comment by NikolasKomonen** [comment]
@fbricon updated

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/XMLAssert.java [hunk 18]
```diff
@@ -112,6 +112,8 @@ public static void testCompletionFor(XMLLanguageService xmlLanguageService, Stri
 				new CompletionSettings(autoCloseTags), expectedItems);
 	}
 
+
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/XMLAssert.java]: unnecessary changes
